File: Compass/app/src/main/java/com/example/compass/MainActivityViewModel.java
package com.example.compass;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.location.Location;
import android.os.Looper;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

import static android.content.Context.SENSOR_SERVICE;

public class MainActivityViewModel extends ViewModel implements SensorEventListener {

    private SensorManager sensorManager;
    private Context context;
    private Place destination;
    private Place myLocation;
    private float angleValue;

    private MutableLiveData<Float> compassDegree = new MutableLiveData<>();
    private MutableLiveData<Double> distance = new MutableLiveData<>();
    private MutableLiveData<Float> angle = new MutableLiveData<>();
    private MutableLiveData<Float> destinationDegree = new MutableLiveData<>();

    FusedLocationProviderClient fusedLocationProviderClient;
    LocationRequest locationRequest;
    LocationCallback locationCallback = new LocationCallback(){
        @Override
        public void onLocationResult(LocationResult locationResult) {
            if(locationResult == null) return;
            for(Location location : locationResult.getLocations()) {
                myLocation = new Place(location.getLatitude(), location.getLongitude());
                if(destination != null) {
                   distance.postValue(destination.meterDistanceBetweenPoints(myLocation.getLatitude(), myLocation.getLongitude()));
                   angleValue = destination.rotationAngle(myLocation.getLatitude(), myLocation.getLongitude());
                   destinationDegree.postValue(angleValue);
                   angle.postValue(angleValue);
               }
            }
        }
    };


    public void init(Context c){
        this.context = c;
        sensorManager = (SensorManager) context.getSystemService(SENSOR_SERVICE);
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context);
        locationRequest = LocationRequest.create();
        locationRequest.setInterval(2000);
        locationRequest.setFastestInterval(1000);
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    }

    public LiveData<Float> getCompassDegree(){
        return compassDegree;
    }

    public LiveData<Double> getDistance(){
        return distance;
    }

    public LiveData<Float> getDestinationDegree(){
        return destinationDegree;
    }

    public LiveData<Float> getAngle(){
        return angle;
    }

    public void setDestination(double lat, double lng){
        destination = new Place(lat, lng);
        if(myLocation != null){
            distance.postValue(destination.meterDistanceBetweenPoints(myLocation.getLatitude(), myLocation.getLongitude()));
            angleValue = destination.rotationAngle(myLocation.getLatitude(), myLocation.getLongitude());
            destinationDegree.postValue(angleValue);
            angle.postValue(angleValue);
        }
    }

    public void registerSensor(){
        sensorManager.registerListener(this, sensorManager.getDefaultSensor(Sensor.TYPE_ORIENTATION), SensorManager.SENSOR_DELAY_FASTEST);
    }

    public void unRegisterSensor(){
        sensorManager.unregisterListener(this);
    }

    @Override
    public void onSensorChanged(SensorEvent event) {
        compassDegree.postValue((float)Math.round(event.values[0]));
        if(destination != null) {
            destinationDegree.postValue((float)Math.round(event.values[0]) + angleValue);
        }
    }

    @Override
    public void onAccuracyChanged(Sensor sensor, int accuracy) {
    }

    public void setLocationUpdates(){
        LocationSettingsRequest request = new LocationSettingsRequest.Builder().addLocationRequest(locationRequest).build();
        SettingsClient client = LocationServices.getSettingsClient(context);
        Task<LocationSettingsResponse> locationSettingsResponseTask = client.checkLocationSettings(request);
        locationSettingsResponseTask.addOnSuccessListener(new OnSuccessListener<LocationSettingsResponse>() {
            @Override
            public void onSuccess(LocationSettingsResponse locationSettingsResponse) {
                startLocationUpdates();
            }
        });

        locationSettingsResponseTask.addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(@NonNull Exception e) {
                e.printStackTrace();
            }
        });
    }

    @SuppressLint("MissingPermission")
    public void startLocationUpdates(){
        fusedLocationProviderClient.requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper());
    }

    public void stopLocationUpdates(){
        fusedLocationProviderClient.removeLocationUpdates(locationCallback);
    }

    public void turnGPSOn(){
        Intent callGPSSettingIntent = new Intent(android.provider.Settings.ACTION_LOCATION_SOURCE_SETTINGS);
        context.startActivity(callGPSSettingIntent);
        setLocationUpdates();
    }

}

File: Compass/app/src/main/java/com/example/compass/MainActivityViewModel.kt
package com.example.compass

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Looper
import android.provider.Settings
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.google.android.gms.location.*

class MainActivityViewModel : ViewModel(), SensorEventListener {
    private var sensorManager: SensorManager? = null
    private var context: Context? = null
    private var destination: Place? = null
    private var myLocation: Place? = null
    private var angleValue = 0f
    
    private val compassDegree = MutableLiveData<Float>()
    private val distance = MutableLiveData<Double>()
    private val angle = MutableLiveData<Float>()
    private val destinationDegree = MutableLiveData<Float>()

    private var fusedLocationProviderClient: FusedLocationProviderClient? = null
    private var locationRequest: LocationRequest? = null

    private var locationCallback: LocationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            if (locationResult == null) return
            for (location in locationResult.locations) {
                myLocation = Place(location.latitude, location.longitude)
                if (destination != null) {
                    distance.postValue(destination!!.meterDistanceBetweenPoints(myLocation!!.getLatitude(), myLocation!!.getLongitude()))
                    angleValue = destination!!.rotationAngle(myLocation!!.getLatitude(), myLocation!!.getLongitude())
                    destinationDegree.postValue(angleValue)
                    angle.postValue(angleValue)
                }
            }
        }
    }

    fun init(c: Context?) {
        context = c
        sensorManager = context!!.getSystemService(Context.SENSOR_SERVICE) as SensorManager
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context!!)
        locationRequest = LocationRequest.create().setInterval(2000).setFastestInterval(1000).setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
    }

    fun getCompassDegree(): LiveData<Float> {
        return compassDegree
    }

    fun getDistance(): LiveData<Double> {
        return distance
    }

    fun getDestinationDegree(): LiveData<Float> {
        return destinationDegree
    }

    fun getAngle(): LiveData<Float> {
        return angle
    }

    fun setDestination(lat: Double, lng: Double) {
        destination = Place(lat, lng)
        if (myLocation != null) {
            distance.postValue(destination!!.meterDistanceBetweenPoints(myLocation!!.getLatitude(), myLocation!!.getLongitude()))
            angleValue = destination!!.rotationAngle(myLocation!!.getLatitude(), myLocation!!.getLongitude())
            destinationDegree.postValue(angleValue)
            angle.postValue(angleValue)
        }
    }

    fun registerSensor() {
        sensorManager!!.registerListener(this, sensorManager!!.getDefaultSensor(Sensor.TYPE_ORIENTATION), SensorManager.SENSOR_DELAY_FASTEST)
    }

    fun unRegisterSensor() {
        sensorManager!!.unregisterListener(this)
    }

    override fun onSensorChanged(event: SensorEvent) {
        compassDegree.postValue(Math.round(event.values[0]).toFloat())
        if (destination != null) {
            destinationDegree.postValue(Math.round(event.values[0]).toFloat() + angleValue)
        }
    }

    override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
    fun setLocationUpdates() {
        val request = LocationSettingsRequest.Builder().addLocationRequest(locationRequest!!).build()
        val client = LocationServices.getSettingsClient(context!!)
        val locationSettingsResponseTask = client.checkLocationSettings(request)
        locationSettingsResponseTask.addOnSuccessListener { startLocationUpdates() }
        locationSettingsResponseTask.addOnFailureListener { e -> e.printStackTrace() }
    }

    @SuppressLint("MissingPermission")
    fun startLocationUpdates() {
        fusedLocationProviderClient!!.requestLocationUpdates(locationRequest, locationCallback, Looper.getMainLooper())
    }

    fun stopLocationUpdates() {
        fusedLocationProviderClient!!.removeLocationUpdates(locationCallback)
    }

    fun turnGPSOn() {
        val callGPSSettingIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
        context!!.startActivity(callGPSSettingIntent)
        setLocationUpdates()
    }
}